torch/testing/_internal/common_utils.py

```diff
@@ -366,13 +366,26 @@ def clear_tracked_input():
 # Wraps an iterator and tracks the most recent value the iterator produces
 # for debugging purposes. Tracked values are stored on the test function.
 class TrackedInputIter:
-    def __init__(self, child_iter, input_type_desc,
-                 callback=lambda x: x, set_seed=True, restrict_to_index=None):
+    def __init__(
+        self,
+        child_iter,
+        input_type_desc,
+        item_callback=None,
+        track_callback=None,
+        set_seed=True,
+        restrict_to_index=None
+    ):
         self.child_iter = enumerate(child_iter)
         # Input type describes the things we're tracking (e.g. "sample input", "error input").
         self.input_type_desc = input_type_desc
-        # Callback is run on each iterated thing to get the thing to track.
-        self.callback = callback
+        # Item callback is run on each (iterated thing, index) to get the thing to return.
```

Comment:
Can you write a comment for why we have to separate the callback and what the expected use case for each would be?

Reply:
Added the comment. It's a bit annoying that this is the case, but the use case is `ErrorInput`, which contains a `SampleInput`. For this case, we want the iterator to return an `ErrorInput` (or a tuple of `ErrorInput` and `subtest_ctx`) but we want the `SampleInput` it contains to be the thing tracked.